crates/lsp/src/lib.rs

```diff
@@ -142,22 +157,94 @@ fn on_did_close_text_document(
     ControlFlow::Continue(())
 }
 
+fn on_did_save_text_document(
+    state: &mut LspState,
+    params: DidSaveTextDocumentParams,
+) -> ControlFlow<Result<(), async_lsp::Error>> {
+    // TODO: Requiring `Language` and `is_opcode_supported` to construct a driver makes for some real stinky code
+    // The driver should not require knowledge of the backend; instead should be implemented as an independent pass (in nargo?)
+    let mut driver = Driver::new(&Language::R1CS, Box::new(|_op| false));
+
+    let file_path = &params.text_document.uri.to_file_path().unwrap();
+
+    driver.create_local_crate(file_path, CrateType::Binary);
+
+    let mut diagnostics = Vec::new();
+
+    if let Err(file_diagnostics) = driver.check_crate() {
+        let fm = driver.file_manager();
+        let files = fm.as_simple_files();
+
+        for FileDiagnostic { file_id, diagnostic } in file_diagnostics {
+            // TODO: This file_id never be 0 because the "path" where it maps is the directory, not a file
+            if file_id.as_usize() != 0 {
+                continue;
+            }
+
+            let mut range = Range::default();
+
+            // TODO: Should this be the first item in secondaries? Should we bail when we find a range?
```

Comment:
I don't understand the purpose of `secondaries` enough to answer this. If we only ever have one secondary, then perhaps we should change the data structure.